oak-commons/src/main/java/org/apache/jackrabbit/oak/commons/sort/ExternalSortByteArray.java

```diff
@@ -0,0 +1,219 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.jackrabbit.oak.commons.sort;
+
+import org.apache.jackrabbit.oak.commons.Compression;
+
+import java.io.BufferedOutputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.EOFException;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.Comparator;
+import java.util.List;
+import java.util.PriorityQueue;
+import java.util.function.Function;
+
+/**
+ * Variation of ExternalSort that stores the lines read from intermediate files as byte arrays to avoid the conversion
+ * from byte[] to String and then back.
+ */
+public class ExternalSortByteArray {
+    public static <T> void mergeSortedFilesBinary(List<Path> files, BufferedOutputStream fbw, final Comparator<T> cmp,
+                                                  boolean distinct, Compression algorithm,
+                                                  Function<T, byte[]> typeToByteArray, Function<byte[], T> byteArrayToType)
+            throws IOException {
+        ArrayList<BinaryFileBuffer<T>> bfbs = new ArrayList<>();
+        try {
+            for (Path f : files) {
+                InputStream in = algorithm.getInputStream(Files.newInputStream(f));
+                BinaryFileBuffer<T> bfb = new BinaryFileBuffer<>(in, byteArrayToType);
+                bfbs.add(bfb);
+            }
+            mergeBinary(fbw, cmp, distinct, bfbs, typeToByteArray);
+        } finally {
+            for (BinaryFileBuffer<T> buffer : bfbs) {
+                try {
+                    buffer.close();
+                } catch (Exception ignored) {
+                }
+            }
+            for (Path f : files) {
+                Files.deleteIfExists(f);
+            }
+        }
+    }
+
+    public static <T> int mergeBinary(BufferedOutputStream fbw, final Comparator<T> cmp, boolean distinct,
+                                      List<BinaryFileBuffer<T>> buffers, Function<T, byte[]> typeToByteArray)
+            throws IOException {
+        PriorityQueue<BinaryFileBuffer<T>> pq = new PriorityQueue<>(
+                11,
+                (i, j) -> cmp.compare(i.peek(), j.peek())
+        );
+        for (BinaryFileBuffer<T> bfb : buffers) {
+            if (!bfb.empty()) {
+                pq.add(bfb);
+            }
+        }
+        int rowcounter = 0;
+        T lastLine = null;
+        try (fbw) {
```

Comment:
the BufferedOutputStream `fwb` is closed twice: here and where it gets created (PipelineMergeSortTask.sortStoreFile#line 314). I think we can remove it from here.

```suggestion
        try {
```

Reply:
Done.